fs/expose/dokan/libdokan.py

```diff
@@ -93,154 +102,166 @@ class DOKAN_FILE_INFO(Structure):
+        LPCWSTR,      # FileName
+        PDOKAN_FILE_INFO)),
+        ("CloseFile", WINFUNCTYPE(None,
+        LPCWSTR,      # FileName
+        PDOKAN_FILE_INFO)),
+        ("ReadFile", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        LPVOID,   # Buffer
+        DWORD,    # NumberOfBytesToRead
+        LPDWORD,  # NumberOfBytesRead
+        LONGLONG, # Offset
+        PDOKAN_FILE_INFO)),
+        ("WriteFile", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        LPCVOID,  # Buffer
+        DWORD,    # NumberOfBytesToWrite
+        LPDWORD,  # NumberOfBytesWritten
+        LONGLONG, # Offset
+        PDOKAN_FILE_INFO)),
+        ("FlushFileBuffers", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        PDOKAN_FILE_INFO)),
+        ("GetFileInformation", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        POINTER(BY_HANDLE_FILE_INFORMATION), # Buffer
+        PDOKAN_FILE_INFO)),
+        ("FindFiles", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # PathName
+        PFillFindData, # call this function with PWIN32_FIND_DATAW
+        PDOKAN_FILE_INFO)),
+        ("FindFilesWithPattern", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # PathName
+        LPCWSTR,  # SearchPattern
+        PFillFindData,  #call this function with PWIN32_FIND_DATAW
+        PDOKAN_FILE_INFO)),
+        ("SetFileAttributes", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR, # FileName
+        DWORD,   # FileAttributes
+        PDOKAN_FILE_INFO)),
+        ("SetFileTime", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR, # FileName
+        POINTER(FILETIME), # CreationTime
+        POINTER(FILETIME), # LastAccessTime
+        POINTER(FILETIME), # LastWriteTime
+        PDOKAN_FILE_INFO)),
+        ("DeleteFile", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR, # FileName
+        PDOKAN_FILE_INFO)),
+        ("DeleteDirectory", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR, # FileName
+        PDOKAN_FILE_INFO)),
+        ("MoveFile", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR, # ExistingFileName
+        LPCWSTR, # NewFileName
+        BOOL,    # ReplaceExisiting
+        PDOKAN_FILE_INFO)),
+        ("SetEndOfFile", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        LONGLONG, # Length
+        PDOKAN_FILE_INFO)),
+        ("SetAllocationSize", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        LONGLONG, # Length
+        PDOKAN_FILE_INFO)),
+        ("LockFile", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        LONGLONG, # ByteOffset
+        LONGLONG, # Length
+        PDOKAN_FILE_INFO)),
+        ("UnlockFile", WINFUNCTYPE(NTSTATUS,
+        LPCWSTR,  # FileName
+        LONGLONG, # ByteOffset
+        LONGLONG, # Length
+        PDOKAN_FILE_INFO)),
+        ("GetDiskFreeSpace", WINFUNCTYPE(NTSTATUS,
+        PULONGLONG, # FreeBytesAvailable
+        PULONGLONG, # TotalNumberOfBytes
+        PULONGLONG, # TotalNumberOfFreeBytes
+        PDOKAN_FILE_INFO)),
+        ("GetVolumeInformation", WINFUNCTYPE(NTSTATUS,
+        LPWSTR,  # VolumeNameBuffer
```

Comment:
Apparently, when invoking Python callbacks, `ctypes` converts all string pointers in parameters to plain Python strings, which will not do for out-parameters. The easiest way to solve this is replace `LPWSTR` with `PVOID`.

This is important, because `ctypes.memove` (which is used in `GetVolumeInformation` callback) works with Python strings without exceptions, but can crash the whole process.